### PR TITLE
feat: tool-output compression crate

### PR DESCRIPTION
**Pull request by brianluby**
## Summary

- Implements the `compression` crate that compresses large tool outputs to ~500 tokens while preserving semantically important content per tool type
- Specialized compressors for Read (language constructs), Bash (errors/warnings/success), Grep (file grouping), Glob (directory grouping), Edit/Write (path + preview), with generic head/tail fallback
- Infallible API with panic recovery via `catch_unwind`, final budget enforcement via truncation
- Includes full spec artifacts (spec, PRD, architecture review, security review, plan, tasks)

## Code Review Fixes Applied

All findings from CodeRabbit review addressed:

| Priority | Fix |
|----------|-----|
| **H-1** | `serde_json` for robust glob JSON parsing (handles commas in paths) |
| **H-2** | `BTreeMap` + secondary sort for deterministic output |
| **H-3** | Documented generic compressor budget coupling |
| **M-1** | Fixed double char iteration in edit compressor |
| **M-2** | Fixed O(n^2) benchmark content generation |
| **M-3** | `debug_assert!` for config validation |
| **M-4** | Epsilon-based `PartialEq` for f64 fields |
| **M-5** | `Display` impl on `ToolType` |
| **M-6** | Documented `cfg(test)` divergence |
| **L-1** | Stronger grep file-path heuristic |
| **L-2** | Exit code patterns in bash error detection |
| **L-6** | SC-006 latency assertion test (<5ms/10K chars) |

## Test plan

- [x] `cargo test` — 295 tests passing (unit, integration, acceptance, smoke)
- [x] `cargo clippy -- -D warnings` — clean (pedantic)
- [x] `cargo fmt --check` — clean
- [ ] `cargo bench` — verify SC-006 latency <5ms per 10K-char input
- [ ] Manual: verify compressed output is readable and semantically useful

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Introduced a tool-output compression system with per-tool compressors, a public compress API, configurable thresholds, and guaranteed output budget enforcement.

* **Documentation**
  * Added comprehensive spec, PRD, architecture review, quickstart, research, security, plan, data model, contracts, and checklists describing behavior and acceptance criteria.

* **Tests**
  * Added extensive unit, integration, smoke, dispatcher, success-criteria tests and benchmark suites validating correctness, budgets, performance, and edge cases.

### CODE REVIEW

File: crates/compression/src/glob.rs [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        if let Ok(paths) = serde_json::from_str::<Vec<String>>(trimmed) {
+            return paths;
+        }
+        // Fallback: manual split for non-standard JSON-like arrays
+        let inner = &trimmed[1..trimmed.len() - 1];
+        return inner
+            .split(',')
+            .map(|s| s.trim().trim_matches('"').to_string())
+            .filter(|s| !s.is_empty())
+            .collect();
```
[the remark this answers is not in the document]
Comment (brianluby) [reply crates/compression/src/glob.rs]: This was already addressed as part of CodeRabbit's review (nitpick on glob.rs lines 77-101). The `parse_paths` docstring now explicitly documents the comma-split fallback limitation and its trade-offs — see lines 77-83 in the current code. Resolving as duplicate.

File: crates/compression/src/lib.rs [hunk 2]
```diff
@@ -1,9 +1,181 @@
...
+
+    let result = panic::catch_unwind(move || {
+        let ctx = context_owned.as_deref();
+        match tool_type_clone {
+            Read => read::compress(&config_clone, &output_owned, ctx),
+            Bash => bash::compress(&config_clone, &output_owned, ctx),
+            Grep => grep::compress(&config_clone, &output_owned, ctx),
+            Glob => glob::compress(&config_clone, &output_owned, ctx),
+            Edit => edit::compress(&config_clone, &output_owned, ctx, false),
+            Write => edit::compress(&config_clone, &output_owned, ctx, true),
```
[the remark this answers is not in the document]
Comment (brianluby) [reply crates/compression/src/lib.rs]: Fixed: replaced cloning with `panic::AssertUnwindSafe` to capture references. The dispatch closure now borrows `config`, `output`, and `input_context` directly, eliminating the O(n) allocations on every compression call.

File: crates/compression/tests/success_criteria.rs [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+        let _result = compress(&config, tool, input, *ctx);
+        let elapsed = start.elapsed();
+
+        assert!(
+            elapsed.as_millis() < margin_ms,
+            "SC-006: {tool} took {}ms (limit: {margin_ms}ms) for {size}-char input",
+            elapsed.as_millis()
+        );
+    }
+}
```
[the remark this answers is not in the document]
Comment (brianluby) [reply crates/compression/tests/success_criteria.rs:383]: Fixed: added `#[ignore = "wall-clock latency assertion is environment-dependent; use cargo bench for SC-006"]` so `cargo test` stays reliable. The Criterion benchmark suite (`benches/compress_bench.rs`) remains the primary SC-006 validation path.

File: crates/compression/src/bash.rs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+use std::fmt::Write;
+use std::sync::LazyLock;
+
+use regex::Regex;
+
+use crate::config::CompressionConfig;
+
+static ERROR_PATTERN: LazyLock<Regex> = LazyLock::new(|| {
+    Regex::new(r"(?i)(?:error[:\[]|ERR!|FAILED|panic(?:ked)?|fatal|cannot find|not found|undefined|segfault|abort|exit(?:ed)?\s+(?:with\s+)?(?:status|code)\s+[1-9])").expect("BUG: invalid regex literal")
+});
```
[the remark this answers is not in the document]
Comment (brianluby) [reply crates/compression/src/bash.rs:16]: Fixed: introduced `regex_util::compile()` helper that logs `tracing::warn` and returns a never-matching fallback regex (`\\z.`) on compilation failure. All `LazyLock<Regex>` initializers in `bash.rs` and `lang.rs` now use this infallible helper instead of `.expect()`.

File: crates/compression/src/lang.rs [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+static RS_FUNCTION: LazyLock<Regex> = LazyLock::new(|| {
+    Regex::new(r"^(?:pub(?:\(crate\))?\s+)?(?:async\s+)?fn\s+\w+")
+        .expect("BUG: invalid regex literal")
+});
+static RS_STRUCT: LazyLock<Regex> = LazyLock::new(|| {
+    Regex::new(r"^(?:pub(?:\(crate\))?\s+)?(?:struct|enum|trait)\s+\w+")
+        .expect("BUG: invalid regex literal")
+});
+static RS_IMPL: LazyLock<Regex> =
+    LazyLock::new(|| Regex::new(r"^impl(?:<[^>]*>)?\s+").expect("BUG: invalid regex literal"));
```
[the remark this answers is not in the document]
Comment (brianluby) [reply crates/compression/src/lang.rs]: Fixed alongside the `bash.rs` fix: all 13 regex patterns in `lang.rs` now use the shared `regex_util::compile()` infallible helper. See the bash.rs thread for details.
